activerecord/lib/active_record/relation/batches.rb

```diff
@@ -222,45 +222,53 @@ def in_batches(of: 1000, start: nil, finish: nil, load: false, error_on_ignore:
       end
 
       relation = relation.reorder(batch_order(order)).limit(batch_limit)
-      relation = apply_limits(relation, start, finish, order)
-      relation.skip_query_cache! # Retaining the results in the query cache would undermine the point of batching
+      relation = apply_finish_limit(relation, finish, order) if finish
       batch_relation = relation
 
       loop do
+        batch_relation = apply_start_limit(relation, start, order) if start
+
         if load
-          records = batch_relation.records
-          ids = records.map(&:id)
-          yielded_relation = where(primary_key => ids)
-          yielded_relation.load_records(records)
+          records = batch_relation.uncached do
+            batch_relation.limit(batch_limit + 1).records
+          end
+
+          start = records[batch_limit]&.id
+          records = records.take(batch_limit)
+
+          break if records.empty?
+
+          raise ArgumentError.new("Primary key not included in the custom select clause") unless records.first.id
+
+          batch_relation.load_records(records)
         else
-          ids = batch_relation.pluck(primary_key)
-          yielded_relation = where(primary_key => ids)
-        end
+          stop = batch_relation.uncached do
+            batch_relation.offset(batch_limit).pick(primary_key)
+          end
 
-        break if ids.empty?
+          if stop
+            batch_relation = apply_finish_limit(batch_relation, stop, order, inclusive: false)
```

Comment:
I think we can use a range here:

```suggestion
            batch_relation = relation.where(primary_key => start..stop)
```

That should significantly simplify the code.

Reply:
Unfortunately, we can't use ranges here, because it generates `id BETWEEN x AND y` sql query, which is equivalent to `id >= x AND id <= y`. When we batching in a `desc` order, `y <= x` and `BETWEEN` would return no results (for example, `x = 10`, `y = 1` - `id >= 10 AND id <= 1` makes no sense).

Reply:
```ruby
order == :asc ? start..stop : stop..start
```

no?

Reply:
This works, but we already have `apply_finish_limit`, used in other place, so this might be confusing why we do not use it.